src/keri/core/kraming.py

```diff
@@ -0,0 +1,231 @@
+from keri import kering
+from keri.help import helping
+
+from typing import NamedTuple
+from keri import help
+from hio.base import Doer
+import json
+
+logger = help.ogler.getLogger()
+
+# TODO: Implement message type as part of TimelinessCache module for more granular control
+class MessageType(NamedTuple):
+    """Named tuple for KERI message types, not yet used."""
+    QRY: str = "qry"
+    RPY: str = "rpy"
+    PRO: str = "pro"
+    BAR: str = "bar"
+    EXN: str = "exn"
+
+
+MESSAGE_TYPES = MessageType()
+
+class TimelinessCache:
+    """TimelinessCache is responsible for preventing replay attacks in KERI/KRAM by ensuring
+    messages are timely and processed in a strictly monotonically ordered fashion.
+
+    It maintains:
+    1. A Lagging Window Size Table - to determine validity windows for different message types
+    2. A Replay Cache Table - to store timestamps of previously validated messages
+    """
+
+    def __init__(self, db, defaultWindowSize=3.0, defaultDriftSkew=1.0):
+        """Initialize the TimelinessCache.
+
+        Parameters:
+            db: Database instance that contains the IoSetSuber at db.time
+            defaultWindowSize (float): Default window size in seconds
+            defaultDriftSkew (float): Default drift skew in seconds
+        """
+        self.defaultWindowSize = defaultWindowSize
+        self.defaultDriftSkew = defaultDriftSkew
+
+        self.db = db
+
+    def setWindowParameters(self, aid, windowSize=None, driftSkew=None, messageType=None):
+        """Set window parameters for given autonomic identifier and message type.
+        """
+        # TODO: Implement message type as part of the window parameters
+        windowSize = windowSize or self.defaultWindowSize
+        driftSkew = driftSkew or self.defaultDriftSkew
+
+        # Serialize the tuple as JSON bytes for storage
+        windowTuple = (windowSize, driftSkew)
+        serialized = json.dumps(windowTuple).encode('utf-8')
+
+        self.db.kram.pin(aid, [serialized])
```

Comment:
If you `.pin()` with the keys of `(aid, messageType)` for a key that would look like "aid.messageType" then you could support window parameters by message type.